subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/ResolvedCoordinates.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.publish.internal.mapping;
+
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents the coordinates that a declared dependency should be published to.
+ * Similar to {@link ModuleVersionIdentifier}, but allows a null version.
+ */
+public interface ResolvedCoordinates {
```

Comment:
💭 I know I said it before, but I still think this should be 1 simple immutable data class with 2 constructors instead of 3 types, 2 of them anonymous.

Reply:
The subclasses of the interface are hidden behind the public static factory methods. Users of this interface do not even need to be aware of the multiple implementations. 

I think this is just fine the way it is. Since this is the second time you commented on this, why do you think this is such a big deal?

Reply:
It's just that the distinct anonymous implementations are unnecessary - just having 1 named type should be the preferred default.  I think you should have to argue that the usefulness of the the anonymous types outweighs their additional complexity vs. the other way around, and I don't think it does here.